farcasterapi/warpcast/warpcast.go

```diff
@@ -0,0 +1,186 @@
+package warpcast
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/google/uuid"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/util"
+)
+
+const (
+	maxRetries            = 12              // Maximum number of retries
+	baseDelay             = 1 * time.Second // Initial delay, increases exponentially
+	WarpcastApi           = "https://api.warpcast.com/v2"
+	warpcastDirectMessage = WarpcastApi + "/ext-send-direct-cast"
+
+	unauthorizedMessage = "Must provide Authorization header with API key"
+)
+
+type dmRequest struct {
+	RecipientFID   uint64 `json:"recipientFid"`
+	Message        string `json:"message"`
+	IdempotencyKey string `json:"idempotencyKey"`
+}
+
+type errResponse struct {
+	Errors []struct {
+		Message string `json:"message"`
+	} `json:"errors"`
+}
+
+type WarpcastAPI struct {
+	userFID      uint64
+	apiKey       string
+	reqSemaphore chan struct{} // Semaphore to limit concurrent requests
+}
+
+func (w *WarpcastAPI) SetFarcasterUser(fid uint64, apiKey string) error {
```

Comment:
I think it would be better to have a constructor NewWarpcastAPI with fid and apiKey instead of the Setter.